### DISCUSSION TITLE
commander.send_setpoint(...) which values/ units in abs and velocity mode

### DISCUSSION THREAD

**Discussion by clubbymaty** [bitcraze/discussions#363] [Q&A]
Hello,

what values in which units does the commander with the send.setpoint accept?

from https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/commander/ i cannot extract what this command expects in absolute mode for the thrust - the Crazyflie just goes wildly upwards if I type in the thrustBase (20000), what leads me to this question.

Where can I see all the steps/ parameter settings etc. the CF Firmware does when calling the commander.send_setpoint(...) function? What are the basic prerequisites (parameter wise) the velocity controller needs to work fine?

Side-Question: What exactly is (+-)- and x-mode?

best regards!

**Comment by ataffanel** [comment]
By default the ```send_sepoint()``` function takes absolute angles for roll/pitch, angular velocity for yaw and thrust power for thrust. The thrust unit can be seen in percent of the available power with 0 being 0% and 65535 being 100%. The Crazyflie cannot be controlled in position or velocity with these parameter, a control loop is required (be it a program or a human). 

There is parameters that allows to enable the onboard position control loop and change the meaning of the parameters, typically mapping roll/pitch/thrust to x/y/z for velocity or position control but these parameter are depreciated and dedicated functions should be used instead. The Crazyflie has a generic commander port that implement a [lot of common control modes](https://github.com/bitcraze/crazyflie-firmware/blob/d89b96226ddd4540171770a036a375c8606b4f1f/src/modules/src/crtp_commander_generic.c#L402-L411) and allows to add more easily. These packet are mapped to python functions in the ```commander``` module.

So I would suggest using the function that corresponds to your use-case instead. For example if you need position control you can use ```send_position_setpoint(x, y, z, yaw)```.

As for the side question: + and X modes are two different possible orientation for a quadcopter: Crazyflie 1 was in + mode which means that one of the motor was facing front, Crazyflie 2.x is in X mode which means that the front direction is between 2 motors.
